protected.go

```diff
@@ -21,31 +21,35 @@ var (
 )
 
 const (
-	dnsServer      = "127.0.0.1"
-	connectTimeOut = 15 * time.Second
-	readDeadline   = 15 * time.Second
-	writeDeadline  = 15 * time.Second
-	socketError    = -1
-	dnsPort        = 7300
+	readDeadline  = 15 * time.Second
+	writeDeadline = 15 * time.Second
+	socketError   = -1
 )
 
+// The actual function to protect a connection. Same signature as
+// https://developer.android.com/reference/android/net/VpnService.html#protect(java.net.Socket)
 type Protect func(fileDescriptor int) error
 
-type ProtectedConn struct {
-	net.Conn
-	mutex    sync.Mutex
-	isClosed bool
-	socketFd int
-	ip       [4]byte
-	port     int
-}
-
 type Protector struct {
 	protect Protect
+	dnsAddr syscall.Sockaddr
+	dns     string
 }
 
-func New(protect Protect) *Protector {
-	return &Protector{protect}
+// New construct a protector from the protect function and DNS server address in host:port format.
+func New(protect Protect, dnsServerAddr string) (*Protector, error) {
```

Comment:
A few common conversion and error checking code are moved from `resolve` to here